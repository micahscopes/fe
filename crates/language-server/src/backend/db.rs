use common::{impl_db_traits, InputDb};

use hir::{HirDb, LowerHirDb, SpannedHirDb};
use hir_analysis::{diagnostics::SpannedHirAnalysisDb, HirAnalysisDb};

use super::get_core::init_core_ingot;
// xxx use salsa::{ParallelDatabase, Snapshot};

#[salsa::db]
pub trait LanguageServerDb:
    salsa::Database + SpannedHirAnalysisDb + HirAnalysisDb + HirDb + LowerHirDb + SpannedHirDb + InputDb
{
}

#[salsa::db]
impl<DB> LanguageServerDb for DB where
    DB: Sized
        + salsa::Database
        + SpannedHirAnalysisDb
        + HirAnalysisDb
        + HirDb
        + LowerHirDb
        + SpannedHirDb
        + InputDb
{
}

#[salsa::db]
#[derive(Clone)]
pub struct LanguageServerDatabase {
    storage: salsa::Storage<Self>,
}

impl Default for LanguageServerDatabase {
    fn default() -> Self {
        let mut db = LanguageServerDatabase {
            storage: Default::default(),
        };

        init_core_ingot(&mut db);
        db
    }
}

impl_db_traits!(
    LanguageServerDatabase,
    InputDb,
    HirDb,
    LowerHirDb,
    SpannedHirDb,
    HirAnalysisDb,
    SpannedHirAnalysisDb,
);
