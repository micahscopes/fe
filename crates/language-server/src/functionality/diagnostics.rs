use std::ops::Range;

use camino::Utf8Path;

use codespan_reporting as cs;
use cs::{diagnostic as cs_diag, files as cs_files};

use common::{
    diagnostics::{CompleteDiagnostic, LabelStyle, Severity},
    InputDb, InputFile,
};

use fxhash::FxHashMap;
use hir::{
    analysis_pass::AnalysisPassManager, diagnostics::DiagnosticVoucher, hir_def::TopLevelMod,
    lower::map_file_to_mod, ParsingPass,
};
use hir_analysis::name_resolution::{
    DefConflictAnalysisPass, ImportAnalysisPass, PathAnalysisPass,
};

use crate::{
    backend::db::{LanguageServerDatabase, LanguageServerDb},
    util::diag_to_lsp,
};

pub trait ToCsDiag {
    fn to_cs(&self, db: &LanguageServerDatabase) -> cs_diag::Diagnostic<InputFile>;
}

impl<T> ToCsDiag for T
where
    T: DiagnosticVoucher + Sync,
{
    fn to_cs(&self, db: &LanguageServerDatabase) -> cs_diag::Diagnostic<InputFile> {
        let complete = self.to_complete(db);

        let severity = convert_severity(complete.severity);
        let code = Some(complete.error_code.to_string());
        let message = complete.message;

        let labels = complete
            .sub_diagnostics
            .into_iter()
            .filter_map(|sub_diag| {
                let span = sub_diag.span?;
                match sub_diag.style {
                    LabelStyle::Primary => {
                        cs_diag::Label::new(cs_diag::LabelStyle::Primary, span.file, span.range)
                    }
                    LabelStyle::Secondary => {
                        cs_diag::Label::new(cs_diag::LabelStyle::Secondary, span.file, span.range)
                    }
                }
                .with_message(sub_diag.message)
                .into()
            })
            .collect();

        cs_diag::Diagnostic {
            severity,
            code,
            message,
            labels,
            notes: vec![],
        }
    }
}

fn convert_severity(severity: Severity) -> cs_diag::Severity {
    match severity {
        Severity::Error => cs_diag::Severity::Error,
        Severity::Warning => cs_diag::Severity::Warning,
        Severity::Note => cs_diag::Severity::Note,
    }
}

#[salsa::tracked(return_ref)]
pub fn file_line_starts(db: &dyn LanguageServerDb, file: InputFile) -> Vec<usize> {
    cs::files::line_starts(file.text(db.as_input_db())).collect()
}

impl<'a> cs_files::Files<'a> for LanguageServerDatabase {
    type FileId = InputFile;
    type Name = &'a Utf8Path;
    type Source = &'a str;

    fn name(&'a self, file_id: Self::FileId) -> Result<Self::Name, cs_files::Error> {
        Ok(file_id.path(self).as_path())
    }

    fn source(&'a self, file_id: Self::FileId) -> Result<Self::Source, cs_files::Error> {
        Ok(file_id.text(self))
    }

    fn line_index(
        &'a self,
        file_id: Self::FileId,
        byte_index: usize,
    ) -> Result<usize, cs_files::Error> {
        let starts = file_line_starts(self, file_id);
        Ok(starts
            .binary_search(&byte_index)
            .unwrap_or_else(|next_line| next_line - 1))
    }

    fn line_range(
        &'a self,
        file_id: Self::FileId,
        line_index: usize,
    ) -> Result<Range<usize>, cs_files::Error> {
        let line_starts = file_line_starts(self, file_id);

        let start = *line_starts
            .get(line_index)
            .ok_or(cs_files::Error::LineTooLarge {
                given: line_index,
                max: line_starts.len() - 1,
            })?;

        let end = if line_index == line_starts.len() - 1 {
            file_id.text(self).len()
        } else {
            *line_starts
                .get(line_index + 1)
                .ok_or(cs_files::Error::LineTooLarge {
                    given: line_index,
                    max: line_starts.len() - 1,
                })?
        };

        Ok(Range { start, end })
    }
}

impl LanguageServerDatabase {
    pub fn analyze_top_mod(&self, top_mod: TopLevelMod) -> Vec<Box<dyn DiagnosticVoucher>> {
        let mut pass_manager = initialize_analysis_pass(self);
        pass_manager.run_on_module(top_mod)
    }

    pub fn finalize_diags(&self, diags: &[Box<dyn DiagnosticVoucher>]) -> Vec<CompleteDiagnostic> {
        let mut diags: Vec<_> = diags.iter().map(|d| d.to_complete(self)).collect();
        diags.sort_by(|lhs, rhs| match lhs.error_code.cmp(&rhs.error_code) {
            std::cmp::Ordering::Equal => lhs.primary_span().cmp(&rhs.primary_span()),
            ord => ord,
        });
        diags
    }

    pub fn get_lsp_diagnostics(
        &self,
        files: Vec<InputFile>,
    ) -> FxHashMap<async_lsp::lsp_types::Url, Vec<async_lsp::lsp_types::Diagnostic>> {
        let mut result =
            FxHashMap::<async_lsp::lsp_types::Url, Vec<async_lsp::lsp_types::Diagnostic>>::default(
            );
        files
            .iter()
            .flat_map(|file| {
                let top_mod = map_file_to_mod(self, *file);
                let diagnostics = self.analyze_top_mod(top_mod);
                self.finalize_diags(&diagnostics)
                    .into_iter()
                    .flat_map(|diag| diag_to_lsp(diag, self.as_input_db()).clone())
            })
            .for_each(|(uri, more_diags)| {
                let _ = result.entry(uri.clone()).or_insert_with(Vec::new);
                let diags = result.entry(uri).or_insert_with(Vec::new);
                diags.extend(more_diags);
            });
        result
    }
}

fn initialize_analysis_pass(db: &LanguageServerDatabase) -> AnalysisPassManager<'_> {
    let mut pass_manager = AnalysisPassManager::new();
    pass_manager.add_module_pass(Box::new(ParsingPass::new(db)));
    pass_manager.add_module_pass(Box::new(DefConflictAnalysisPass::new(db)));
    pass_manager.add_module_pass(Box::new(ImportAnalysisPass::new(db)));
    pass_manager.add_module_pass(Box::new(PathAnalysisPass::new(db)));
    pass_manager
}
