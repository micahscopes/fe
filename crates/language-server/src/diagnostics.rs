use std::ops::Range;

use camino::Utf8Path;
use codespan_reporting as cs;
use cs::{diagnostic as cs_diag, files as cs_files};

use common::{
    diagnostics::{LabelStyle, Severity},
    InputFile,
};
use hir::diagnostics::DiagnosticVoucher;

use crate::db::{LanguageServerDataBase, LanguageServerDb};


pub trait ToCsDiag {
    fn to_cs(&self, db: &LanguageServerDataBase) -> cs_diag::Diagnostic<InputFile>;
}

impl<T> ToCsDiag for T
where
    T: DiagnosticVoucher,
{
    fn to_cs(&self, db: &LanguageServerDataBase) -> cs_diag::Diagnostic<InputFile> {
        let complete = self.to_complete(db);

        let severity = convert_severity(complete.severity);
        let code = Some(complete.error_code.to_string());
        let message = complete.message;

        let labels = complete
            .sub_diagnostics
            .into_iter()
            .filter_map(|sub_diag| {
                let span = sub_diag.span?;
                match sub_diag.style {
                    LabelStyle::Primary => {
                        cs_diag::Label::new(cs_diag::LabelStyle::Primary, span.file, span.range)
                    }
                    LabelStyle::Secondary => {
                        cs_diag::Label::new(cs_diag::LabelStyle::Secondary, span.file, span.range)
                    }
                }
                .with_message(sub_diag.message)
                .into()
            })
            .collect();

        cs_diag::Diagnostic {
            severity,
            code,
            message,
            labels,
            notes: vec![],
        }
    }
}

fn convert_severity(severity: Severity) -> cs_diag::Severity {
    match severity {
        Severity::Error => cs_diag::Severity::Error,
        Severity::Warning => cs_diag::Severity::Warning,
        Severity::Note => cs_diag::Severity::Note,
    }
}

#[salsa::tracked(return_ref)]
pub fn file_line_starts(db: &dyn LanguageServerDb, file: InputFile) -> Vec<usize> {
    cs::files::line_starts(file.text(db.as_input_db())).collect()
}

impl<'a> cs_files::Files<'a> for LanguageServerDataBase {
    type FileId = InputFile;
    type Name = &'a Utf8Path;
    type Source = &'a str;

    fn name(&'a self, file_id: Self::FileId) -> Result<Self::Name, cs_files::Error> {
        Ok(file_id.path(self).as_path())
    }

    fn source(&'a self, file_id: Self::FileId) -> Result<Self::Source, cs_files::Error> {
        Ok(file_id.text(self))
    }

    fn line_index(
        &'a self,
        file_id: Self::FileId,
        byte_index: usize,
    ) -> Result<usize, cs_files::Error> {
        let starts = file_line_starts(self, file_id);
        Ok(starts
            .binary_search(&byte_index)
            .unwrap_or_else(|next_line| next_line - 1))
    }

    fn line_range(
        &'a self,
        file_id: Self::FileId,
        line_index: usize,
    ) -> Result<Range<usize>, cs_files::Error> {
        let line_starts = file_line_starts(self, file_id);

        let start = *line_starts
            .get(line_index)
            .ok_or(cs_files::Error::LineTooLarge {
                given: line_index,
                max: line_starts.len() - 1,
            })?;

        let end = if line_index == line_starts.len() - 1 {
            file_id.text(self).len()
        } else {
            *line_starts
                .get(line_index + 1)
                .ok_or(cs_files::Error::LineTooLarge {
                    given: line_index,
                    max: line_starts.len() - 1,
                })?
        };

        Ok(Range { start, end })
    }
}
