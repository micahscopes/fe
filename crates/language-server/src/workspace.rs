use std::{
    collections::BTreeSet,
    path::{Path, PathBuf},
};

use common::{
    input::{IngotKind, Version},
    InputFile, InputIngot,
};
use patricia_tree::StringPatriciaMap;

use crate::db::LanguageServerDatabase;

const FE_CONFIG_SUFFIX: &str = "fe.toml";

trait IngotFileContext {
    fn get_input_for_file_path(&self, path: &str) -> Option<InputFile>;
    fn get_ingot_for_file_path(&self, path: &str) -> Option<InputIngot>;
}

struct Ingot {
    local: InputIngot,
    external: InputIngot,
}

pub struct LocalIngotContext {
    db: LanguageServerDatabase,
    // cache `InputIngot` for path
    ingot: InputIngot,
    external_ingots: StringPatriciaMap<InputIngot>,
    // cache `InputFile` for path
    files: StringPatriciaMap<InputFile>,
}

fn ingot_contains_file(ingot_path: &str, file_path: &str) -> bool {
    let ingot_path = ingot_path
        .strip_suffix(&FE_CONFIG_SUFFIX)
        .unwrap_or(ingot_path);
    file_path.starts_with(ingot_path)
}

fn get_containing_ingot<'a, T>(ingots: &'a StringPatriciaMap<T>, path: &'a str) -> Option<T> {
    ingots
        .get_longest_common_prefix(path)
        .filter(|(ingot_path, _)| ingot_contains_file(ingot_path, path))
        .map(|(_, ingot)| *ingot)
}

impl LocalIngotContext {
    pub fn new(db: LanguageServerDatabase, config_path: &str) -> Option<Self> {
        let ingot = InputIngot::new(
            &db,
            config_path,
            IngotKind::Local,
            Version::new(0, 0, 0),
            BTreeSet::new(),
        );
        Some(Self {
            db,
            ingot,
            external_ingots: StringPatriciaMap::new(),
            files: StringPatriciaMap::new(),
        })
    }
}

impl IngotFileContext for LocalIngotContext {
    fn get_input_for_file_path(&self, path: &str) -> Option<InputFile> {
        self.files.get(path).map_or_else(
            || {
                let ingot = self.get_ingot_for_file_path(path)?;
                let file = InputFile::new(&self.db, ingot, path.into(), "".into());
                self.files.insert(path.to_string(), file.clone());
                Some(file)
            },
            |file| Some(*file),
        )
    }

    fn get_ingot_for_file_path(&self, path: &str) -> Option<InputIngot> {
        get_containing_ingot(&self.external_ingots, path).or_else(|| Some(self.ingot.clone()))
    }
}

struct StandaloneIngotContext {
    db: LanguageServerDatabase,
    ingots: StringPatriciaMap<InputIngot>,
    files: StringPatriciaMap<InputFile>,
}

impl StandaloneIngotContext {
    pub fn new(db: LanguageServerDatabase) -> Self {
        Self {
            db,
            ingots: StringPatriciaMap::new(),
            files: StringPatriciaMap::new(),
        }
    }
}

impl IngotFileContext for StandaloneIngotContext {
    fn get_input_for_file_path(&self, path: &str) -> Option<InputFile> {
        self.files.get(path).map_or_else(
            || {
                let ingot = self.get_ingot_for_file_path(path)?;
                let file = InputFile::new(&self.db, ingot, path.into(), "".into());
                self.files.insert(path.to_string(), file.clone());
                Some(file)
            },
            |file| Some(*file),
        )
    }

    fn get_ingot_for_file_path(&self, path: &str) -> Option<InputIngot> {
        get_containing_ingot(&self.ingots, path)
    }
}

struct Workspace {
    db: LanguageServerDatabase,
    ingot_contexts: StringPatriciaMap<LocalIngotContext>,
    standalone_ingot_contexts: StandaloneIngotContext,
}

impl Workspace {
    pub fn new(db: LanguageServerDatabase) -> Self {
        Self {
            db,
            ingot_contexts: StringPatriciaMap::new(),
            standalone_ingot_contexts: StandaloneIngotContext::new(db),
        }
    }

    pub fn get_ingot_context(&self, config_path: &str) -> Option<LocalIngotContext> {
        self.ingot_contexts.get(config_path).map_or_else(
            || {
                let ingot = InputIngot::new(
                    &self.db,
                    config_path,
                    IngotKind::Local,
                    Version::new(0, 0, 0),
                    BTreeSet::new(),
                );
                let context = LocalIngotContext::new(self.db, config_path);
                self.ingot_contexts
                    .insert(config_path.to_string(), context.unwrap());
                context
            },
            |context| Some(*context),
        )
    }

    pub fn get_ingot_for_file_path(&self, path: &str) -> Option<InputIngot> {
        get_containing_ingot(&self.ingot_contexts, path).map_or_else(
            || self.standalone_ingot_contexts.get_ingot_for_file_path(path),
            |ingot| Some(ingot.ingot),
        )
    }

    pub fn get_input_file_for_file_path(&self, path: &str) -> Option<InputFile> {
        self.get_ingot_for_file_path(path)
            .map_or_else(|| None, |ingot| Some(ingot.root_file(&self.db)))
    }
}
