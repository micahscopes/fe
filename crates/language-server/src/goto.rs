use fxhash::FxHashMap;
use hir::{
    hir_def::{scope_graph::ScopeId, ItemKind, PathId, TopLevelMod},
    visitor::{prelude::LazyPathSpan, Visitor, VisitorCtxt},
    HirDb,
};
use hir_analysis::name_resolution::EarlyResolvedPath;

use crate::db::{LanguageServerDatabase, LanguageServerDb};
use common::diagnostics::Span;
use hir::span::LazySpan;

pub(crate) type GotoEnclosingPath = (PathId, ScopeId);
pub(crate) type GotoPathMap = FxHashMap<Span, GotoEnclosingPath>;

pub struct PathSpanCollector<'db> {
    path_map: GotoPathMap,
    db: &'db dyn LanguageServerDb,
}

impl<'db> PathSpanCollector<'db> {
    pub fn new(db: &'db LanguageServerDatabase) -> Self {
        Self {
            path_map: FxHashMap::default(),
            db,
        }
    }
}

pub(crate) type Cursor = rowan::TextSize;

impl<'db> Visitor for PathSpanCollector<'db> {
    fn visit_path(&mut self, ctxt: &mut VisitorCtxt<'_, LazyPathSpan>, path: PathId) {
        let Some(span) = ctxt
            .span()
            .map(|lazy_span| lazy_span.resolve(self.db.as_spanned_hir_db()))
            .flatten()
        else {
            return;
        };

        let scope = ctxt.scope();
        self.path_map.insert(span, (path, scope));
    }
}

fn smallest_enclosing_path(cursor: Cursor, path_map: &GotoPathMap) -> Option<GotoEnclosingPath> {
    let mut smallest_enclosing_path = None;
    let mut smallest_range_size = None;

    for (span, enclosing_path) in path_map {
        if span.range.contains(cursor) {
            let range_size = span.range.end() - span.range.start();
            if smallest_range_size.is_none() || range_size < smallest_range_size.unwrap() {
                smallest_enclosing_path = Some(*enclosing_path);
                smallest_range_size = Some(range_size);
            }
        }
    }

    return smallest_enclosing_path;
}

pub fn goto_enclosing_path(
    db: &mut LanguageServerDatabase,
    top_mod: TopLevelMod,
    cursor: Cursor,
) -> Option<EarlyResolvedPath> {
    // Find the innermost item enclosing the cursor.
    let item: ItemKind = db.find_enclosing_item(top_mod, cursor)?;

    let mut visitor_ctxt = VisitorCtxt::with_item(db.as_hir_db(), item);
    let mut path_collector = PathSpanCollector::new(&db);
    path_collector.visit_item(&mut visitor_ctxt, item);

    let path_map = path_collector.path_map;

    // Find the path that encloses the cursor.
    let goto_starting_path = smallest_enclosing_path(cursor, &path_map)?;

    let (path_id, scope_id) = goto_starting_path;

    // Resolve path.
    let resolved_path = hir_analysis::name_resolution::resolve_path_early(db, path_id, scope_id);

    Some(resolved_path)
}

#[cfg(test)]
mod tests {
    use crate::workspace::{IngotFileContext, Workspace};

    use super::*;
    use common::input::IngotKind;
    use dir_test::{dir_test, Fixture};
    use fe_compiler_test_utils::snap_test;
    use std::path::Path;

    fn extract_multiple_cursor_positions_from_spans(
        db: &mut LanguageServerDatabase,
        top_mod: TopLevelMod,
    ) -> Vec<rowan::TextSize> {
        let mut visitor_ctxt = VisitorCtxt::with_top_mod(db.as_hir_db(), top_mod);
        let mut path_collector = PathSpanCollector::new(&db);
        path_collector.visit_top_mod(&mut visitor_ctxt, top_mod);

        let path_map = path_collector.path_map;

        let mut cursors = Vec::new();
        for (span, _) in path_map {
            let cursor = span.range.start();
            // println!("cursor from span: {:?}, {:?}", span, cursor);
            cursors.push(cursor);
        }

        cursors
    }

    #[dir_test(
        dir: "$CARGO_MANIFEST_DIR/test_files/single_ingot",
        glob: "**/lib.fe",
    )]
    fn test_goto_multiple_files(fixture: Fixture<&str>) {
        let cargo_manifest_dir = std::env::var("CARGO_MANIFEST_DIR").unwrap();
        let ingot_base_dir = Path::new(&cargo_manifest_dir).join("test_files/single_ingot");

        let db = &mut LanguageServerDatabase::default();
        let workspace = &mut Workspace::default();

        let _ = workspace.set_workspace_root(db, &Some(ingot_base_dir.clone()));

        let fe_source_path = ingot_base_dir.join(fixture.path());
        let input = workspace.input_from_file_path(db, fixture.path());
        assert_eq!(input.unwrap().ingot(db).kind(db), IngotKind::Local);

        let top_mod = workspace.top_mod_from_file(db, &fe_source_path, fixture.content()).unwrap();

        let ingot = workspace.ingot_from_file_path(db, fixture.path());
        assert_eq!(ingot.unwrap().kind(db), IngotKind::Local);

        let cursors = extract_multiple_cursor_positions_from_spans(db, top_mod);
        let mut cursor_path_map: FxHashMap<Cursor, String> = FxHashMap::default();

        cursors.iter().for_each(|cursor| {
            let early_resolution = goto_enclosing_path(db, top_mod, *cursor);

            let goto_info = match early_resolution {
                Some(EarlyResolvedPath::Full(bucket)) => {
                    if bucket.len() > 0 {
                        bucket
                            .iter()
                            .map(|x| x.pretty_path(db).unwrap())
                            .collect::<Vec<_>>()
                            .join("\n")
                    } else {
                        String::from("`NameResBucket` is empty")
                    }
                }
                Some(EarlyResolvedPath::Partial {
                    res,
                    unresolved_from: _,
                }) => res.pretty_path(db).unwrap(),
                None => String::from("No resolution available"),
            };

            cursor_path_map.insert(*cursor, goto_info);
        });

        let result = format!(
            "{}\n---\n{}",
            fixture.content(),
            cursor_path_map
                .iter()
                .map(|(cursor, path)| {
                    format!("cursor position: {:?}, path: {:?}", cursor, path)
                })
                .collect::<Vec<_>>()
                .join("\n")
        );
        snap_test!(result, fixture.path());
    }

    #[dir_test(
        dir: "$CARGO_MANIFEST_DIR/test_files",
        glob: "goto*.fe"
    )]
    fn test_goto_enclosing_path(fixture: Fixture<&str>) {
        let mut db = &mut LanguageServerDatabase::default();
        let workspace = &mut Workspace::default();
        let path = Path::new(fixture.path());
        let top_mod = workspace.top_mod_from_file(&mut db, path, fixture.content()).unwrap();

        let cursors = extract_multiple_cursor_positions_from_spans(&mut db, top_mod);

        let mut cursor_path_map: FxHashMap<Cursor, String> = FxHashMap::default();

        cursors.iter().for_each(|cursor| {
            let resolved_path = goto_enclosing_path(db, top_mod, *cursor);

            match resolved_path {
                Some(path) => match path {
                    EarlyResolvedPath::Full(bucket) => {
                        let path = bucket
                            .iter()
                            .map(|x| x.pretty_path(db).unwrap())
                            .collect::<Vec<_>>()
                            .join("\n");
                        cursor_path_map.insert(*cursor, path);
                    }
                    EarlyResolvedPath::Partial {
                        res,
                        unresolved_from: _,
                    } => {
                        let path = res.pretty_path(db).unwrap();
                        cursor_path_map.insert(*cursor, path);
                    }
                },
                None => {}
            };
        });

        let result = format!(
            "{}\n---\n{}",
            fixture.content(),
            cursor_path_map
                .iter()
                .map(|(cursor, path)| { format!("cursor position: {:?}, path: {}", cursor, path) })
                .collect::<Vec<_>>()
                .join("\n")
        );
        snap_test!(result, fixture.path());
    }

    #[dir_test(
        dir: "$CARGO_MANIFEST_DIR/test_files",
        glob: "smallest_enclosing*.fe"
    )]
    fn test_smallest_enclosing_path(fixture: Fixture<&str>) {
        let db = &mut LanguageServerDatabase::default();
        let workspace = &mut Workspace::default();
        let path = Path::new(fixture.path());
        let top_mod = workspace.top_mod_from_file(db, path, fixture.content()).unwrap();

        let cursors = extract_multiple_cursor_positions_from_spans(db, top_mod);

        let mut cursor_path_map: FxHashMap<Cursor, String> = FxHashMap::default();

        cursors.iter().for_each(|cursor| {
            let mut visitor_ctxt = VisitorCtxt::with_top_mod(db.as_hir_db(), top_mod);
            let mut path_collector = PathSpanCollector::new(&db);
            path_collector.visit_top_mod(&mut visitor_ctxt, top_mod);

            let path_map = path_collector.path_map;
            let enclosing_path = smallest_enclosing_path(*cursor, &path_map);

            let resolved_enclosing_path = hir_analysis::name_resolution::resolve_path_early(
                db,
                enclosing_path.unwrap().0,
                enclosing_path.unwrap().1,
            );

            let res = match resolved_enclosing_path {
                EarlyResolvedPath::Full(bucket) => bucket
                    .iter()
                    .map(|x| x.pretty_path(db).unwrap())
                    .collect::<Vec<_>>()
                    .join("\n"),
                EarlyResolvedPath::Partial {
                    res,
                    unresolved_from: _,
                } => res.pretty_path(db).unwrap(),
            };
            cursor_path_map.insert(*cursor, res);
        });

        let result = format!(
            "{}\n---\n{}",
            fixture.content(),
            cursor_path_map
                .iter()
                .map(|(cursor, path)| { format!("cursor position: {:?}, path: {}", cursor, path) })
                .collect::<Vec<_>>()
                .join("\n")
        );
        snap_test!(result, fixture.path());
    }
}
