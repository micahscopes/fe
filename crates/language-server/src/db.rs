use std::{collections::BTreeSet, path};

use common::{
    diagnostics::CompleteDiagnostic,
    input::{IngotKind, Version},
    InputDb, InputFile, InputIngot,
};
use hir::{
    analysis_pass::AnalysisPassManager, diagnostics::DiagnosticVoucher, hir_def::{TopLevelMod, ItemKind},
    lower::map_file_to_mod, HirDb, LowerHirDb, ParsingPass, SpannedHirDb, span::{DynLazySpan, LazySpan},
};
use hir_analysis::{
    name_resolution::{DefConflictAnalysisPass, ImportAnalysisPass, PathAnalysisPass},
    HirAnalysisDb,
};

use crate::goto::Cursor;

#[salsa::jar(db = LanguageServerDb)]
pub struct Jar(crate::diagnostics::file_line_starts);

pub trait LanguageServerDb:
    salsa::DbWithJar<Jar> + HirAnalysisDb + HirDb + LowerHirDb + SpannedHirDb + InputDb
{ }

impl<DB> LanguageServerDb for DB where
    DB: Sized + salsa::DbWithJar<Jar> + HirAnalysisDb + HirDb + LowerHirDb + SpannedHirDb + InputDb
{ }


#[salsa::db(common::Jar, hir::Jar, hir::LowerJar, hir::SpannedJar, hir_analysis::Jar, Jar)]
pub struct LanguageServerDatabase {
    storage: salsa::Storage<Self>,
    diags: Vec<Box<dyn DiagnosticVoucher>>,
}

impl LanguageServerDatabase {
    pub fn run_on_top_mod(&mut self, top_mod: TopLevelMod) {
        self.run_on_file_with_pass_manager(top_mod, initialize_analysis_pass);
    }

    pub fn run_on_file_with_pass_manager<F>(&mut self, top_mod: TopLevelMod, pm_builder: F)
    where
        F: FnOnce(&LanguageServerDatabase) -> AnalysisPassManager<'_>,
    {
        self.diags.clear();
        self.diags = {
            let mut pass_manager = pm_builder(self);
            pass_manager.run_on_module(top_mod)
        };
    }

    pub fn top_mod_from_file(&mut self, file_path: &path::Path, source: &str) -> TopLevelMod {
        let kind = IngotKind::StandAlone;

        // We set the ingot version to 0.0.0 for stand-alone file.
        let version = Version::new(0, 0, 0);
        let root_file = file_path;
        let ingot = InputIngot::new(
            self,
            file_path.parent().unwrap().as_os_str().to_str().unwrap(),
            kind,
            version,
            BTreeSet::new(),
        );

        let file_name = root_file.file_name().unwrap().to_str().unwrap();
        let file = InputFile::new(self, ingot, file_name.into(), source.to_string());
        ingot.set_root_file(self, file);
        ingot.set_files(self, [file].into());

        map_file_to_mod(self, file)
    }

    pub fn find_enclosing_item(&mut self, top_mod: TopLevelMod, cursor: Cursor) -> Option<ItemKind> {
        let items = top_mod.scope_graph(self.as_hir_db()).items_dfs(self.as_hir_db());

        let mut smallest_enclosing_item = None;
        let mut smallest_range_size = None;

        for item in items {
            let lazy_item_span = DynLazySpan::from(item.lazy_span());
            let item_span = lazy_item_span.resolve(SpannedHirDb::as_spanned_hir_db(self)).unwrap();

            if item_span.range.contains(cursor) {
                let range_size = item_span.range.end() - item_span.range.start();
                if smallest_range_size.is_none() || range_size < smallest_range_size.unwrap() {
                    smallest_enclosing_item = Some(item);
                    smallest_range_size = Some(range_size);
                }
            }
        }

        return smallest_enclosing_item;
    }
    
    pub fn finalize_diags(&self) -> Vec<CompleteDiagnostic> {
        let mut diags: Vec<_> = self.diags.iter().map(|d| d.to_complete(self)).collect();
        diags.sort_by(|lhs, rhs| match lhs.error_code.cmp(&rhs.error_code) {
            std::cmp::Ordering::Equal => lhs.primary_span().cmp(&rhs.primary_span()),
            ord => ord,
        });
        diags
    }
}

impl salsa::Database for LanguageServerDatabase {
    fn salsa_event(&self, _: salsa::Event) {}
}

impl Default for LanguageServerDatabase {
    fn default() -> Self {
        let db = Self {
            storage: Default::default(),
            diags: Vec::new(),
        };
        db.prefill();
        db
    }
}

fn initialize_analysis_pass(db: &LanguageServerDatabase) -> AnalysisPassManager<'_> {
    let mut pass_manager = AnalysisPassManager::new();
    pass_manager.add_module_pass(Box::new(ParsingPass::new(db)));
    pass_manager.add_module_pass(Box::new(DefConflictAnalysisPass::new(db)));
    pass_manager.add_module_pass(Box::new(ImportAnalysisPass::new(db)));
    pass_manager.add_module_pass(Box::new(PathAnalysisPass::new(db)));
    pass_manager
}
