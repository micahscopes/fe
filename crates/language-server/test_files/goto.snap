---
source: crates/language-server/src/goto.rs
assertion_line: 153
expression: result
input_file: crates/language-server/test_files/goto.fe
---
struct Foo {}
struct Bar {}

fn main() {
    let x: Foo
    let y: Bar
    let z: baz::Baz
}

mod baz {
   pub struct Baz {}
}
---
cursor position: 64, path: 
cursor position: 82, path: goto::baz::Baz
cursor position: 79, path: 
cursor position: 52, path: goto::Foo
cursor position: 49, path: 
cursor position: 67, path: goto::Bar
